### PR TITLE
Don't await the countDocuments() promise COMPASS-5279

### PR DESCRIPTION
**Pull request by lerouxb**
Quite proud of my little `waitForStates()` test helper.

### CODE REVIEW

File: packages/compass-crud/src/stores/crud-store.js [hunk 1]
```diff
@@ -201,6 +202,7 @@ const configureStore = (options = {}) => {
         isReadonly: false,
         isTimeSeries: false,
         status: DOCUMENTS_STATUS_INITIAL,
+        debouncing: false,
```
Comment (lerouxb) [inline packages/compass-crud/src/stores/crud-store.js:205]: This is not a very good variable name for what's going on,  but I can't think of anything better yet.

File: packages/compass-crud/src/stores/crud-store.spec.js [hunk 2]
```diff
@@ -20,62 +20,120 @@ const CONNECTION = new Connection({
   mongodb_database_name: 'admin'
 });
 
-function listenToStore(store, cb, expectedCalls = 1) {
-  return new Promise((resolve, reject) => {
-    let numCalls = 0;
-    const unsubscribe = store.listen((state) => {
-      ++numCalls;
+function delay(timeout) {
```
Comment (lerouxb) [inline packages/compass-crud/src/stores/crud-store.spec.js:23]: I have written this function so many times but I cannot bring myself to make a leftpad package for it.

File: packages/compass-crud/src/stores/crud-store.js [hunk 3]
```diff
@@ -398,7 +400,7 @@ const configureStore = (options = {}) => {
             this.state.docs.splice(index, 1);
             this.setState({
               count: this.state.count - 1,
-              end: this.state.end - 1
+              end: Math.max(this.state.end - 1, 0)
```
Comment (lerouxb) [inline packages/compass-crud/src/stores/crud-store.js:403]: drive-by

File: packages/compass-crud/src/stores/crud-store.spec.js (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-crud/src/stores/crud-store.spec.js:23]: It exists, but only in node 15+ https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options

**Review by gribnoysup — APPROVED** [review]
Weird that we don't have a loading state for the count and it just shows N/A, but at least the UI is not broken and the queries are fast now

File: packages/compass-crud/src/stores/crud-store.spec.js (hunk 2 above)
Comment (addaleax) [reply packages/compass-crud/src/stores/crud-store.spec.js:23]: There’s also `util.promisify(setTimeout)`, fwiw, which does the same thing (but obviously still requires some extra code for importing util).

Comment (gribnoysup) [reply packages/compass-crud/src/stores/crud-store.spec.js:23]: Woah, didn't know that `setTimeout` implements a custom handler for promisify, neat!

Comment (lerouxb) [reply packages/compass-crud/src/stores/crud-store.spec.js:23]: I have such déjà vu right now. Like I've belatedly discovered this before..

File: packages/compass-components/src/components/spin-loader.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  return render(<SpinLoader size="12px" />);
+}
+
+describe('SpinLoader Component', function () {
+  afterEach(function () {
+    cleanup();
+  });
+
+  it('should show the spinner', function () {
+    renderLoader();
```
Comment (lerouxb) [inline packages/compass-components/src/components/spin-loader.spec.tsx:17]: I have no idea how we'd test this. What's there to check for? It has no text, no sub-elements, no interactivity.. :)

File: packages/compass-shell/src/components/shell-header/shell-header.jsx [hunk 5]
```diff
@@ -41,7 +41,10 @@ export class ShellHeader extends Component {
             &gt;_MONGOSH
             {!isExpanded && isOperationInProgress && (
               <>
-                <SpinLoader size="12px" />
+                <ShellLoader
```
Comment (lerouxb) [inline packages/compass-shell/src/components/shell-header/shell-header.jsx]: Couldn't get shell-header's tests to pas with ShellLoader in there even though it works when I test manually. Got the feeling it had to do with enzyme and newer components?

**Review by Anemy — COMMENTED** [review]
Looks real good, huge speedup when looking at a larger collection. 
Two thoughts on how we present this two users:
- I was wondering if it would make sense to increase the contrast of the loader on the light background? I thinks it's a bit light since it was built for the dark background of the browser repl/shell.
- Usually it resolves fast enough, but for a user that might see this and be unsure what it's performing, could we add a tooltip that shows `Fetching document count...` or something? I wonder if always showing the `of` in the `Displaying documents 1-20 of 1000`, even when loading, might make the operation in progress more apparent?


https://user-images.githubusercontent.com/1791149/142035161-fc3a4575-0910-46c9-915d-b77dc93bd6ff.mp4
